File: python/3.1.2_TrafficLight.py
#!/usr/bin/env python3

import RPi.GPIO as GPIO
import time
import threading

#define the pins connect to 74HC595
SDI   = 18      #serial data input(DS)
RCLK  = 16      #memory clock input(STCP)
SRCLK = 12       #shift register clock input(SHCP)
number = (0xc0,0xf9,0xa4,0xb0,0x99,0x92,0x82,0xf8,0x80,0x90)

placePin = (11,13,15,19)   
ledPin =(22,24,26) 

greenLight = 30
yellowLight = 5
redLight = 60
lightColor=("Red","Green","Yellow")

colorState=0
counter = 60      
t = 0         


def setup():
    GPIO.setmode(GPIO.BOARD)   
    GPIO.setup(SDI, GPIO.OUT)   
    GPIO.setup(RCLK, GPIO.OUT)
    GPIO.setup(SRCLK, GPIO.OUT)
    for pin in placePin:
        GPIO.setup(pin,GPIO.OUT)
    for pin in ledPin:
        GPIO.setup(pin,GPIO.OUT)
    
            
def hc595_shift(dat):     
	for bit in range(0, 8):	
		GPIO.output(SDI, 0x80 & (dat << bit))
		GPIO.output(SRCLK, GPIO.HIGH)
		GPIO.output(SRCLK, GPIO.LOW)
	GPIO.output(RCLK, GPIO.HIGH)
	GPIO.output(RCLK, GPIO.LOW)

def selectPlace(digit): 
    GPIO.output(placePin[0],GPIO.LOW if (digit&0x08) else GPIO.HIGH)
    GPIO.output(placePin[1],GPIO.LOW if (digit&0x04) else GPIO.HIGH)
    GPIO.output(placePin[2],GPIO.LOW if (digit&0x02) else GPIO.HIGH)
    GPIO.output(placePin[3],GPIO.LOW if (digit&0x01) else GPIO.HIGH)

def display(num):  
    hc595_shift(0xff) 
    selectPlace(0x01)   
    hc595_shift(number[num%10]) 
    time.sleep(0.003)   
    hc595_shift(0xff)
    selectPlace(0x02) 
    hc595_shift(number[num%100//10])
    time.sleep(0.003)
    hc595_shift(0xff)
    selectPlace(0x04)  
    hc595_shift(number[num%1000//100])
    time.sleep(0.003)
    hc595_shift(0xff)
    selectPlace(0x08) 
    hc595_shift(number[num%10000//1000])
    time.sleep(0.003)

def timer():        #timer function
    global counter
    global colorState
    global t
    t = threading.Timer(1.0,timer)  
    t.start()     
    counter-=1                          
    if (counter is 0):
        if(colorState is 0):
            counter= greenLight
        if(colorState is 1):
            counter=yellowLight
        if (colorState is 2):
            counter=redLight
        colorState=(colorState+1)%3
    print ("counter : %d    color: %s "%(counter,lightColor[colorState]))

def lightup(state):
    for i in range(0,3):
        GPIO.output(ledPin[i], GPIO.HIGH)
    GPIO.output(ledPin[state], GPIO.LOW)

def loop():
    global t
    global counter
    global colorState
    t = threading.Timer(1.0,timer)   
    t.start()                          
    while True:
        display(counter)
        lightup(colorState)
        
def destroy():   # When "Ctrl+C" is pressed, the function is executed. 
    global t
    GPIO.cleanup()      
    t.cancel()      #cancel the timer

if __name__ == '__main__': # Program starting from here 
    setup() 
    try:
        loop()  
    except KeyboardInterrupt:  
        destroy()  
 

File: c/3.1.2/3.1.2_TrafficLight.c
/**********************************************************************
* Filename    : 3.1.2_TrafficLight.c
* Description : 
* Author      : Robot
* E-mail      : support@sunfounder.com
* website     : www.sunfounder.com
* Update      : Jimmy  2019/08/06 
**********************************************************************/
#include <wiringPi.h>
#include <stdio.h>
#include <wiringShift.h>
#include <signal.h>
#include <unistd.h>
#define     SDI     5   //serial data input(DS)
#define     RCLK    4   //memory clock input(STCP)
#define     SRCLK    1    //shift register clock input(SHCP)
const int placePin[]={0,2,3,12};        // Define 4 digit's common pin
const int ledPin[]={6,10,11};   //Define 3 LED pin
unsigned char number[]={0xc0,0xf9,0xa4,0xb0,0x99,0x92,0x82,0xf8,0x80,0x90};

int greenLight = 30;
int yellowLight = 5;
int redLight = 60;
int colorState = 0;
char *lightColor[]={"Red","Green","Yellow"};
int counter = 60;

//here is the 4 digital function
void selectPlace(int digit){    
    digitalWrite(placePin[0],!(digit&0x08));
    digitalWrite(placePin[1],!(digit&0x04));
    digitalWrite(placePin[2],!(digit&0x02));
    digitalWrite(placePin[3],!(digit&0x01));
}

void hc595_shift(int8_t data){      //To assign 8 bit value to 74HC595’s shift register
    int i;  
    for(i = 0; i < 8; i++){
        digitalWrite(SDI, 0x80 & (data << i));
        digitalWrite(SRCLK,1);
        delayMicroseconds(1);
        digitalWrite(SRCLK,0);
	}
        digitalWrite(RCLK,1);
        delayMicroseconds(1);
        digitalWrite(RCLK,0);        
}

void display(int num){  //display the number
	hc595_shift(0xff);	
    selectPlace(0x01);      
    hc595_shift(number[num%10]); 
    delay(1); 
 
    hc595_shift(0xff);
    selectPlace(0x02); 
    hc595_shift(number[num%100/10]);
    delay(1);
   
    hc595_shift(0xff);    
    selectPlace(0x04); 
    hc595_shift(number[num%1000/100]);
    delay(1);

    hc595_shift(0xff);    
    selectPlace(0x08);
    hc595_shift(number[num%10000/1000]);
    delay(1);
}

//here is the timer for countdown 
void timer(int  sig){       //Timer function
    if(sig == SIGALRM){   
        counter --;         
        alarm(1); 
        if(counter == 0){
            if(colorState == 0) counter = greenLight;
            if(colorState == 1) counter = yellowLight;
            if(colorState == 2) counter = redLight;
            colorState = (colorState+1)%3; 
        }
        printf("counter : %d \t light color: %s \n",counter,lightColor[colorState]);
    }
}

void lightup(int state)
{
    for(int i=0;i<3;i++){
        digitalWrite(ledPin[i],HIGH);
    }
    digitalWrite(ledPin[state],LOW);    
}

int main(void)
{
    int i;


    if(wiringPiSetup() == -1){ //when initialize wiring failed,print messageto screen
        printf("setup wiringPi failed !");
        return 1; 
    }
    pinMode(SDI,OUTPUT);        //set the pin connected to74HC595 for output mode
    pinMode(RCLK,OUTPUT);
    pinMode(SRCLK,OUTPUT);
    //set the pin connected to 7-segment display common end to output mode
    for(i=0;i<4;i++){       
        pinMode(placePin[i],OUTPUT);
        digitalWrite(placePin[i],HIGH);
    }
    //set the led pin
    for(i=0;i<3;i++){       
        pinMode(ledPin[i],OUTPUT);
        digitalWrite(ledPin[i],HIGH);
    }
    signal(SIGALRM,timer);  //configure the timer
    alarm(1);               //set the time of timer to 1s
    while(1){
        display(counter);   //display the number counter
        lightup(colorState); //turn on the traffic light
    }
    return 0;
}